### ISSUE TITLE
Spack tries to load uninstalled dependency

### ISSUE THREAD

**Issue by michaelkuhn** [LLNL/spack#30733]
### Steps to reproduce

I've managed to run into a rather weird bug. I want to install (among others) meson and libfuse:
```
$ spack install meson
# Dependencies get installed
[+] $SPACK/opt/spack/linux-fedora36-zen2/gcc-12.1.1/meson-0.62.1-xaykn6zufjdqcytgq73itpumpt7iu24i

$ spack install libfuse
# No dependencies get installed since everything is available
[+] $SPACK/opt/spack/linux-fedora36-zen2/gcc-12.1.1/libfuse-3.10.5-lhgxftn53ghk5ih654hzhesdv4dfa64s

$ spack gc           
==> The following packages will be uninstalled:

    -- linux-fedora36-zen2 / gcc@12.1.1 -----------------------------
    u2gyyfy berkeley-db@18.1.40  qp5cnif perl@5.34.1    iu4eubw py-pip@21.3.1
    l732qd2 diffutils@3.8        5ocoz6k pkgconf@1.8.0  34hd52u py-wheel@0.37.0

==> Do you want to proceed? [y/N] y
==> Successfully uninstalled pkgconf@1.8.0%gcc@12.1.1 arch=linux-fedora36-zen2/5ocoz6k
==> Successfully uninstalled diffutils@3.8%gcc@12.1.1 arch=linux-fedora36-zen2/l732qd2
==> Successfully uninstalled perl@5.34.1%gcc@12.1.1+cpanm+shared+threads arch=linux-fedora36-zen2/qp5cnif
==> Successfully uninstalled py-pip@21.3.1%gcc@12.1.1 arch=linux-fedora36-zen2/iu4eubw
==> Successfully uninstalled py-wheel@0.37.0%gcc@12.1.1 arch=linux-fedora36-zen2/34hd52u
==> Successfully uninstalled berkeley-db@18.1.40%gcc@12.1.1+cxx~docs+stl patches=b231fcc arch=linux-fedora36-zen2/u2gyyfy

$ spack load libfuse
# OK
```
A few build dependencies were removed and loading libfuse works fine.

However, if I specify the generic `x86_64` target (which we do in CI), things get weird:
```
$ spack config --scope site add 'packages:all:target:[x86_64]'

$ spack install meson                                         
# Dependencies get installed as above
[+] $SPACK/opt/spack/linux-fedora36-x86_64/gcc-12.1.1/meson-0.62.1-johgfvspdfhai4vdz545mwrz5j3fjabi

$ spack install libfuse
# Some dependencies get installed a second time (gdbm, python, meson etc.) 
[+] $SPACK/opt/spack/linux-fedora36-x86_64/gcc-12.1.1/libfuse-3.10.5-xc2kyxejdosydhjw4gh7nkfnxkljjbo2

$ spack gc           
==> The following packages will be uninstalled:

    -- linux-fedora36-x86_64 / gcc@12.1.1 ---------------------------
    gkck3wz berkeley-db@18.1.40  l46g35l perl@5.34.1           mft2ifk py-wheel@0.37.0
    5vxs6ms diffutils@3.8        t5lucpb pkgconf@1.8.0         hk2icd6 py-wheel@0.37.0
    ry4g3o7 gdbm@1.23            krmyayg py-pip@21.3.1         fxvielj python@3.9.12
    ntnz7bu meson@0.62.1         zhumtxo py-pip@21.3.1
    lbee2ny ninja@1.10.2         bcoodrr py-setuptools@59.4.0

==> Do you want to proceed? [y/N] y
==> Successfully uninstalled pkgconf@1.8.0%gcc@12.1.1 arch=linux-fedora36-x86_64/t5lucpb
==> Successfully uninstalled diffutils@3.8%gcc@12.1.1 arch=linux-fedora36-x86_64/5vxs6ms
==> Successfully uninstalled perl@5.34.1%gcc@12.1.1+cpanm+shared+threads arch=linux-fedora36-x86_64/l46g35l
==> Successfully uninstalled py-pip@21.3.1%gcc@12.1.1 arch=linux-fedora36-x86_64/zhumtxo
==> Successfully uninstalled py-wheel@0.37.0%gcc@12.1.1 arch=linux-fedora36-x86_64/mft2ifk
==> Successfully uninstalled py-pip@21.3.1%gcc@12.1.1 arch=linux-fedora36-x86_64/krmyayg
==> Successfully uninstalled py-wheel@0.37.0%gcc@12.1.1 arch=linux-fedora36-x86_64/hk2icd6
==> Successfully uninstalled meson@0.62.1%gcc@12.1.1 patches=aa6c50d arch=linux-fedora36-x86_64/ntnz7bu
==> Successfully uninstalled berkeley-db@18.1.40%gcc@12.1.1+cxx~docs+stl patches=b231fcc arch=linux-fedora36-x86_64/gkck3wz
==> Successfully uninstalled ninja@1.10.2%gcc@12.1.1 arch=linux-fedora36-x86_64/lbee2ny
==> Successfully uninstalled py-setuptools@59.4.0%gcc@12.1.1 arch=linux-fedora36-x86_64/bcoodrr
==> Successfully uninstalled python@3.9.12%gcc@12.1.1+bz2+ctypes+dbm~debug+ensurepip+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93,4c24573,f2fd060 arch=linux-fedora36-x86_64/fxvielj
==> Successfully uninstalled gdbm@1.23%gcc@12.1.1 arch=linux-fedora36-x86_64/ry4g3o7

$ spack load libfuse        
==> Error: list index out of range
```
Notice how the duplicated dependencies got garbage collected since they were pure build dependencies. I tracked down the `list index out of range` error to python's `headers` method, that is, somehow loading libfuse tries to look up the headers of the uninstalled(!) python package, which obviously fails.

This does *not* happen if I install libfuse using `target=x86_64` on the command line.

All of this was done in two freshly cloned repositories.

### Information on your system

* **Spack:** 0.18.0.dev0 (f9aa7c611cfa084a5237381389c9b169f61b094a)
* **Python:** 3.10.4
* **Platform:** linux-fedora36-zen2
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by alalazo** [comment]
I was able to reproduce using:

* **Spack:** 0.18.0.dev0 (3bc656808c6453f1f76353f32885719cfaba4eb4)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

**Comment by alalazo** [comment]
@tgamblin This is one of the "build" inconsistencies we talked about for "full-hash", when we decided how to implement the hack to permit building specs with conflicting build dependencies in multiple steps + a possible bug (or feature) on reusing the root spec. The reproducer can be simplified further:
```console
$ spack install meson
$ spack spec -Il meson 
Input spec
--------------------------------
 -   meson

Concretized
--------------------------------
 -   7eui4ld  meson@0.62.1%gcc@9.4.0 patches=aa6c50d arch=linux-ubuntu20.04-x86_64
 -   cmpd34d      ^ninja@1.10.2%gcc@9.4.0 arch=linux-ubuntu20.04-x86_64
 -   3p7245s          ^python@3.9.12%gcc@9.4.0+bz2+ctypes+dbm~debug+ensurepip+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93,4c24573,f2fd060 arch=linux-ubuntu20.04-x86_64
[+]  ehf5627              ^bzip2@1.0.8%gcc@9.4.0~debug~pic+shared arch=linux-ubuntu20.04-x86_64
[+]  p5ebbuy                  ^diffutils@3.8%gcc@9.4.0 arch=linux-ubuntu20.04-x86_64
[+]  qnzwa5k                      ^libiconv@1.16%gcc@9.4.0 libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]  poaz426              ^expat@2.4.8%gcc@9.4.0+libbsd arch=linux-ubuntu20.04-x86_64
[+]  cv3ofog                  ^libbsd@0.11.5%gcc@9.4.0 arch=linux-ubuntu20.04-x86_64
[+]  deksstr                      ^libmd@1.0.4%gcc@9.4.0 arch=linux-ubuntu20.04-x86_64
 -   jhgm6vn              ^gdbm@1.23%gcc@9.4.0 arch=linux-ubuntu20.04-x86_64
[+]  hse57kg                  ^readline@8.1%gcc@9.4.0 arch=linux-ubuntu20.04-x86_64
[+]  7ofkj5g                      ^ncurses@6.2%gcc@9.4.0~symlinks+termlib abi=none arch=linux-ubuntu20.04-x86_64
[+]  tgufarn                          ^pkgconf@1.8.0%gcc@9.4.0 arch=linux-ubuntu20.04-x86_64
[+]  siczzyk              ^gettext@0.21%gcc@9.4.0+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-ubuntu20.04-x86_64
[+]  6en4s73                  ^libxml2@2.9.13%gcc@9.4.0~python arch=linux-ubuntu20.04-x86_64
[+]  pug5rin                      ^xz@5.2.5%gcc@9.4.0~pic libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]  cvc7wjl                      ^zlib@1.2.12%gcc@9.4.0+optimize+pic+shared patches=0d38234 arch=linux-ubuntu20.04-x86_64
[+]  6jp2whc                  ^tar@1.34%gcc@9.4.0 zip=pigz arch=linux-ubuntu20.04-x86_64
[+]  5v6qbpo                      ^pigz@2.7%gcc@9.4.0 arch=linux-ubuntu20.04-x86_64
[+]  iz2z5mz                      ^zstd@1.5.2%gcc@9.4.0+programs compression=none libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]  dpfg22g              ^libffi@3.4.2%gcc@9.4.0 arch=linux-ubuntu20.04-x86_64
[+]  del3f6o              ^openssl@1.1.1o%gcc@9.4.0~docs~shared certs=system arch=linux-ubuntu20.04-x86_64
[+]  xjse77e                  ^perl@5.34.1%gcc@9.4.0+cpanm+shared+threads arch=linux-ubuntu20.04-x86_64
[+]  bmmadng                      ^berkeley-db@18.1.40%gcc@9.4.0+cxx~docs+stl patches=b231fcc arch=linux-ubuntu20.04-x86_64
[+]  ehf5627                      ^bzip2@1.0.8%gcc@9.4.0~debug~pic+shared arch=linux-ubuntu20.04-x86_64
[+]  p5ebbuy                          ^diffutils@3.8%gcc@9.4.0 arch=linux-ubuntu20.04-x86_64
[+]  qnzwa5k                              ^libiconv@1.16%gcc@9.4.0 libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]  ovdf2pq                      ^gdbm@1.19%gcc@9.4.0 arch=linux-ubuntu20.04-x86_64
[+]  hse57kg                          ^readline@8.1%gcc@9.4.0 arch=linux-ubuntu20.04-x86_64
[+]  7ofkj5g                              ^ncurses@6.2%gcc@9.4.0~symlinks+termlib abi=none arch=linux-ubuntu20.04-x86_64
[+]  tgufarn                                  ^pkgconf@1.8.0%gcc@9.4.0 arch=linux-ubuntu20.04-x86_64
[+]  nam6ocr              ^sqlite@3.38.5%gcc@9.4.0+column_metadata+dynamic_extensions+fts~functions+rtree arch=linux-ubuntu20.04-x86_64
[+]  j4isjhe              ^util-linux-uuid@2.37.4%gcc@9.4.0 arch=linux-ubuntu20.04-x86_64
 -   dajpxjj      ^py-pip@21.3.1%gcc@9.4.0 arch=linux-ubuntu20.04-x86_64
 -   yhzahg4      ^py-setuptools@59.4.0%gcc@9.4.0 arch=linux-ubuntu20.04-x86_64
 -   lx6umrf          ^py-wheel@0.37.0%gcc@9.4.0 arch=linux-ubuntu20.04-x86_64
```

**Comment by alalazo** [comment]
For the solution above I get:
```console
$ spack solve -Il meson
==> Best of 7 considered solutions.
==> Optimization Criteria:
  Priority  Criterion                                            Installed  ToBuild
  1         number of packages to build (vs. reuse)                      -       24
  2         deprecated versions used                                     0        0
  3         version weight                                               0        0
  4         number of non-default variants (roots)                       0        0
  5         preferred providers for roots                                0        0
  6         default values of variants not being used (roots)            0        0
  7         number of non-default variants (non-roots)                   1        0
  8         preferred providers (non-roots)                              0        0
  9         compiler mismatches                                          0        0
  10        OS mismatches                                                0        0
  11        non-preferred OS's                                           0        0
  12        version badness                                              0        0
  13        default values of variants not being used (non-roots)        2        0
  14        non-preferred compilers                                      0        0
  15        target mismatches                                            0        0
  16        non-preferred targets                                      -90     -720
```
It seems to me 24 builds are a bit too much (since they seem 7 from the spec)?

EDIT: 24 is correct. The tree above doesn't show other build dependencies that will be built. Now the question is why do we prefer to build them?

**Comment by alalazo** [comment]
Ok, here's why. We never converted `non-preferred targets` to have a best value of `0`. It can go negative, at which point it is preferable to build all the deps that don't actually make any higher priority criteria to get worse.

EDIT: To convert this we need to review the semantics (which we never thought through fully with the old concretizer) for preferences. We may want to have a table lookup of weights by package (with a default) and add some rule for the target mismatch. A full solution is, I think, not for v0.18 but I wonder if we can have some stopgap.

**Comment by alalazo** [comment]
Fixed by #31266
